crates/kubelet/src/backoff/mod.rs

```diff
@@ -0,0 +1,104 @@
+//! Provides backoff timing control for Kubernetes pod states
+//! such as ImagePullBackoff and CrashLoopBackoff.
+use std::time::Duration;
+
+/// Determines how long to back off before performing a retry.
+#[async_trait::async_trait]
+pub trait BackoffStrategy: Send {
+    /// Resets the strategy after a success.
+    fn reset(&mut self);
+    /// Gets how long to wait before retrying.
+    fn next_duration(&mut self) -> Duration;
+    /// Waits the prescribed amount of time (as per `next_duration`).
+    async fn wait(&mut self) {
+        tokio::time::delay_for(self.next_duration()).await
+    }
+}
+
+/// A `BackoffStrategy` in which the durations increase exponentially
+/// until hitting a cap.
+pub struct ExponentialBackoffStrategy {
+    base_duration: Duration,
+    cap: Duration,
+    last_duration: Duration,
+}
+
+impl ExponentialBackoffStrategy {
+    /// Gets a backoff strategy that adheres to the Kubernetes defaults.
+    pub fn kubernetes_default() -> Self {
+        Self {
+            base_duration: Duration::from_secs(10),
+            cap: Duration::from_secs(300),
+            last_duration: Duration::from_secs(0),
+        }
+    }
+
+    fn capped_next_duration(&self) -> Duration {
+        let next_duration = if self.last_duration == Duration::from_secs(0) {
+            self.base_duration
+        } else {
+            self.last_duration * 2
```

Comment:
Most exponential backoff/retry algorithms I've seen have a jitter component. Should we do that here?

Reply:
I wondered about that. The description of backoff in the k8s docs didn't mention jitter, but I can't imagine anything relies on there _not_ being jitter, and it may indeed be desirable!

Reply:
Honestly up to you. I won't block on it at all. I just know from looking into exponential backoff and the logic behind it that jitter helps avoid problems on retries